ReSharper.FSharp/src/FSharp/FSharp.Psi.Features/src/FSharpTypingAssist.fs

```diff
@@ -736,49 +739,57 @@ type FSharpTypingAssist(lifetime, dependencies) as this =
         let mutable allowedTokens = null
         if not (tryDeindentTokens.TryGetValue(lexer.TokenType, &allowedTokens)) then false else
 
-        let prevTokenEnd =
+        let tokenBeforeKeywordEnd =
             use cookie = LexerStateCookie.Create(lexer)
             lexer.Advance(-1)
             while isIgnored lexer.TokenType do
                 lexer.Advance(-1)
             lexer.TokenEnd
 
-        match x.GetFSharpTree(textControl) with
-        | None -> false
-        | Some parseTree ->
-
-        let document = textControl.Document
-        let visitor =
-            { new SyntaxVisitorBase<_>() with
-                member x.VisitExpr(_, _, defaultTraverse, expr) =
-                    match expr with
-
-                    // if expr then {caret}
-                    // No info is available after error recovery, we check that range is surrounded with if ... then.
-                    | SynExpr.FromParseError (ExprRange range, _)
-
-                    // while expr do {caret}
-                    | SynExpr.While (_, ExprRange range, _, _) when
-                            getEndOffset document range = prevTokenEnd ->
-                        Some range
-
-                    // for i = ... to ... do {caret}
-                    | SynExpr.For (_, _, IdentRange range, _, _, _, _, ExprRange lastRange, _)
+        let expr: IFSharpExpression = getNode textControl lexer.TokenStart
+        if isNull expr then false else
 
-                    // for ... in ... do {caret}
-                    | SynExpr.ForEach (_, _, _, _, PatRange range, ExprRange lastRange, _, _) when
-                            getEndOffset document lastRange = prevTokenEnd ->
-                        Some range
+        let indentOffset =
+            let inline tryGetIndentOffset (exprBeforeKeyword: IFSharpExpression) (indentAnchor: ITreeNode) =
```

Comment:
What is the reason to inline it?

Reply:
In order not to allocate a closure for `tokenBeforeKeywordEnd`, used inside

Reply:
And I also don't see where and how exactly it's allocated. Could you explain, please?

Reply:
It could be viewed in Release mode in IL

![image](https://github.com/user-attachments/assets/4ebcbc5b-ec46-494e-b088-6265f052a749)
![image](https://github.com/user-attachments/assets/e859951d-d457-482e-8fff-a5148f41587b)

In any case, if this is an obstacle on the way to merge, I will remove inline, this is not so critical.

Reply:
Wow, I'd not expect a closure there at all. Thanks for showing this!

Reply:
Maybe it's worth trying to find a minimal repro and reporting it to the compiler team. Or is it expected and I'm missing something?